tsweb/request_id.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package tsweb
+
+import (
+	"context"
+	"net/http"
+)
+
+// RequestID is an opaque identifier for a HTTP request, used to correlate
+// user-visible errors with backend server logs. The RequestID is typically
+// threaded through an HTTP Middleware (WithRequestID) and then can be
+// extracted by HTTP Handlers to include in their logs.
+//
+// RequestID is an opaque identifier for a HTTP request, used to correlate
+// user-visible errors with backend server logs. If present in the context, the
+// RequestID will be printed alongside the message text and logged in the
+// AccessLogRecord.
+//
+// There is no particular format required for a RequestID, but ideally it should
+// be obvious to an end-user that it is something to record for support
+// purposes. One possible example for a RequestID format is:
+// REQ-{server identifier}-{timestamp}-{random hex string}.
+type RequestID string
+
+// RequestIDHeader is a custom HTTP header that the WithRequestID middleware
+// uses to determine whether to re-use a given request ID from the client
+// or generate a new one.
+const RequestIDHeader = "X_TAILSCALE_REQUEST_ID"
+
+// WithRequestID is an http middleware that injects a RequestID in the
+// *http.Request Context. The value of that request id is either retrieved from
+// the RequestIDHeader or from the generator if the header does not exist. Inner
+// handlers can retrieve this ID from the RequestIDFromContext function.
+func WithRequestID(generator func(*http.Request) string) Middleware {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			id := r.Header.Get(RequestIDHeader)
+			if id == "" {
+				id = generator(r)
+			}
+			ctx := SetRequestID(r.Context(), RequestID(id))
+			r = r.WithContext(ctx)
+			h.ServeHTTP(w, r)
+		})
+	}
+}
+
+type requestIDKey struct{}
+
+// RequestIDFromContext retrieves the RequestID from context that can be set by
+// the SetRequestID function.
+func RequestIDFromContext(ctx context.Context) RequestID {
+	val, _ := ctx.Value(requestIDKey{}).(RequestID)
+	return val
+}
+
+// SetRequestID sets the given request id value in the given context.
```

Comment:
Does this need to be exported? I don't think it will be used outside this package, right?

Reply:
@creachadair I think it's worth exporting (as much as I'd like to keep package APIs small) for the following reason: 

In the case of unit tests outside of this package or repo, I'd like to test that a function correctly calls tsweb.RequestIDFromContext, whcih means we'll need to be able to set it in the test first instead of having to go through `WithRequestID` because the unit test will be further down the stack and not explicitly related to HTTP middlewares. 

That said i'm totally happy to skip and un-export it 😁

Reply:
Not a strong objection, the main reason I noticed it is that the `WithX` is more commonly used to attach values to contexts, e.g., `context.WithValue` and other libraries do similarly. Here, though, the `With*` name is already occupied. I considered suggesting swapping them, but if we didn't need it outside the package that wouldn't be a problem.

That said, if you want to export it for testing that seems fine too.

Reply:
@creachadair that's a good point! I ended up un-exporting it for now until I (or someone else) needs it ✌🏼